lib/clerk/authenticate_request.rb

```diff
@@ -0,0 +1,245 @@
+                if auth_context.development_instance?
+                    # TODO: log possible Clock skew detected
+                
+                    # Retry with a generous clock skew allowance (1 day)
+                    claims = verify_token(session_token, timeout: 86_400)
+                    return signed_in(env, claims, session_token) if claims
+                end
+                
+                # Raise error if handshake resolution fails in production
+                raise e
+            end
+        end
+
+        def handle_handshake_maybe_status(env, **opts)
+            return signed_out if !eligible_for_handshake?
+            handshake(env, **opts)
+        end
+
+        # A outcome
+        def handshake(env, **opts)
+            redirect_headers = { LOCATION_HEADER => redirect_to_handshake }
+            [307, debug_auth_headers(**opts).merge(redirect_headers), []]
+        end
+
+        # B outcome
+        def signed_out(**opts)
+            headers = opts.delete(:headers) || {}
+            enforce_auth = opts.delete(:enforce_auth)
+
+            if enforce_auth
+                [401, debug_auth_headers(**opts).merge(headers), []]
+            else
+                [nil, headers, []]
+            end
+        end
+
+        # C outcome
+        def signed_in(env, claims, token, **headers)
+            env["clerk"] = ProxyV2.new(session_claims: claims, session_token: token)
+            [nil, headers, []]
+        end
+
+        def eligible_for_handshake?
+            auth_context.document_request? || (!auth_context.document_request? && auth_context.accepts_html?)
+        end
+
+        private
+
+        def redirect_to_handshake
+            redirect_url = auth_context.clerk_url.dup
+            remove_from_query_string(redirect_url, DEV_BROWSER_COOKIE)
+
+            handshake_url = URI.parse("https://#{auth_context.frontend_api}/v1/client/handshake")
+            handshake_url_qs = Rack::Utils.parse_query(handshake_url.query)
+            handshake_url_qs["redirect_url"] = redirect_url
+
+            if auth_context.development_instance? && auth_context.dev_browser?
+                handshake_url_qs[DEV_BROWSER_COOKIE] = auth_context.dev_browser
+            end
+
+            handshake_url.query = handshake_url_qs.to_query
+            handshake_url.to_s
+        end
+
+        def remove_from_query_string(url, key)
+            qs = Rack::Utils.parse_query(url.query)
+            qs.delete(key)
+            url.query = qs.to_query
+        end
+
+        def verify_token(token, **opts)
+            return false if token.nil? || token.strip.empty?
+        
+            begin
+                sdk.verify_token(token, **opts)
+            rescue JWT::ExpiredSignature, JWT::InvalidIatError => e
+                raise e
+            rescue JWT::DecodeError, JWT::RequiredDependencyError => e
+                false
+            end
+        end
+
+        def sdk
+            Clerk::SDK.new
```

Comment:
:question: Should we memoize this?

Reply:
I thought of doing that in a separate PR. I kept it the same with our current codebase.
@agis Could you remember if there is a reason behind not memorizing this?
Also since we are using a cache for the jwks keys which is shared across the SDK instances it seems okay to leave it as is.
```ruby
# lib/clerk/sdk.rb
@@jwks_cache = JWKSCache.new(JWKS_CACHE_LIFETIME)
```

Reply:
Yep, this is deliberately not memoized, since we don't want it to be shared across requests. See https://github.com/clerk/clerk-sdk-ruby/commit/11239115d79603f5178bfdada9fe94369427818f.